Sources/MongoSwift/Operations/Operation.swift

```diff
@@ -56,29 +73,47 @@ internal class OperationExecutor {
         client: MongoClient,
         session: ClientSession?
     ) -> EventLoopFuture<T.OperationResult> {
-        // early exit and don't attempt to use the thread pool if we've already closed the client.
-        guard !client.isClosed else {
-            return self.makeFailedFuture(MongoClient.ClosedClientError)
+        // don't attempt to use the thread pool if the client is shutting down or shut down.
+        let error: Error? = client.stateLock.withLock {
+            guard client.state == .upAndRunning else {
+                return MongoClient.ShutdownClientError
+            }
+            return nil
+        }
+        if let err = error {
+            return self.makeFailedFuture(err)
         }
 
         // closure containing the work to run in the thread pool: obtain a connection and execute the operation.
         let doOperation = { () -> ExecuteResult<T.OperationResult> in
-            // it's possible that the client was closed in between submitting this task and it being executed, so we
-            // check again here.
-            guard !client.isClosed else {
-                throw MongoClient.ClosedClientError
+            let connResult: Result<Connection?, Error> = client.stateLock.withLock {
```

Comment:
this change basically makes it so we only give you a connection if the client is still open. we were attempting to do this before, but without using the lock there could be an interleaving where the client is closed in between you checking if its closed and actually trying to check out a connection, which could lead to an error when you attempt to check out a connection.

we could just let the connection pool report that error, but it seems kind of racy/weird since it could lead to calls you make after calling `MongoClient.shutdown` going through if the call to shut down the connection pool hasn't been executed yet.

related to this, you can end up with some bad issues if you still have cursors out in the wild when shutdown is called.... the connections' memory leaks because they aren't in the pool at the time `destroy` is called, and when they try to check themselves back in they fail to do so.
we might consider calling `mongoc_client_destroy` directly on the orphaned connection when those cases occur which is not documented as a valid solution but I think would clean them up, but I would also say this is somewhat user error and will only really happen when your application is terminating anyway, so 🤷‍♀

Reply:
Hm yeah, that does seem like a bit of an issue. Maybe we should just document the possibility of a leak for now until we come up with something better. Given client's aren't thread safe, I can't really think of anything though. One way we can combat it is to check for the number of references to the `MongoClient`, and if it's greater than 1 just return an error. Not sure if that would be frustrating for users or not.

Reply:
I don't consider that a leak. It's a positive aspect of our connection pool design that connections are allowed to live on after a pool is closed (in fact this is a feature depended on by the new "quiesce mode" in replication). The connections will be closed/destroyed when they are eventually returned to the pool, either because the operation completed or the cursor was explicitly closed. If a cursor is not exhausted or explicitly closed after a client/pool is closed then I would consider that user error

Reply:
I think in this case it actually is a leak because libmongoc considers destroying clients derived from a pool after the pool is closed to be UB. I may be confusing this with something else though.

Reply:
as it stands, libmongoc's pool as no in-between state where it is closed but not destroyed. if you destroy a pool and there are still checked-out clients, they all become non-functional, because destroying the pool shuts down SDAM. 
consequently, our pool is the same way --`Connection`s have to be checked in before `ConnectionPool.close()` is called.

possibly, we could implement such behavior where connections can live on ourselves, adding some intermediate state where our pool is considered closed and won't let you check out new connections, but the underlying client pool has not yet been destroyed.

then the question becomes: when *do* we call `mongoc_client_pool_destroy`? that method is tricky and cannot simply be called in `deinit` because it needs to be run via the executor, as it does IO (sends `endSessions`.)
if we could come up with a way to have `ConnectionPool.close()` return a future that resolves once all the connections are returned to the pool and the pool has been destroyed, and then shut down the executor once that future has resolved I think it could work...

Reply:
Ah I see, I misunderstood libmongoc's pooling behavior. But still, this is the lock that I am most concerned about. You acquire the lock, and then do a _whole bunch_ of stuff, including potentially reentering this method (an operation can end up calling something that calls another operation, etc).

I think at very least what we want to do reduce the scope of the critical section as much as possible. It seems like you want to be able to "fail fast" if the client is closed, so you should be able to release the lock as soon as you check state, correct?

Reply:
all that happens in the body of `withLock` is:
```swift
guard client.state == .open else {
    return .failure(MongoClient.ClosedClientError)
}

do {
    // select a connection in following order of priority:
    // 1. connection specifically provided for use with this operation
    // 2. if a session was provided, use its underlying connection
    // 3. a new connection from the pool, if available
    return .success(
        try connection ??
        session?.getConnection(forUseWith: client) ??
        client.connectionPool.tryCheckOut()
    )
} catch {
    return .failure(error)
}
```
basically it's the state check, and then also we try to figure out a connection to use / if one is available. my reasoning for doing the connection part with the lock is without it, you could have the following interleaving:
1. we grab the lock here, see client is open, and then release it
2. user calls `MongoClient.close`, client state transitions to `.closing`
3. we let you check out a new connection here, as the operation to close the connection pool hasn't been executed yet, even though we've already marked the client as `.closing`

Reply:
It's incredibly hard to see this with the large comment thread there, but I don't think that's right. From what I can [see](https://github.com/mongodb/mongo-swift-driver/blob/e85695953cb567ff4498954f7b0b3815a01a7266/Sources/MongoSwift/Operations/Operation.swift#L89-L107) it looks like the `withLock` statement contains the `do` statement. What you put in your comment above this one looks correct to me.

Reply:
you don't think your previous statement is right, or that what I said is right? lol 

the do/catch is just the connection logic

Reply:
OK, I see now that the functional change here is that you are putting the lock around just the client state check and connection checkout. 

What's weird about this is that there are paths here where you don't actually try to check out a connection from the pool. How are you handling those cases? (connection is explicitly passed it, or bound to a session) The lock here has a smell.

Reply:
so I'm realizing we should probably just get rid of the explicit connection thing. it isn't used anywhere in the driver anymore.
we used to use it for `NextOperation` just to avoid checking out an extra connection that would go unused (we would pass in the cursor's underlying connection instead), but `NextOperation` still didn't even explicitly use the passed-in cursor connection as the `mongoc_cursor_t` keeps a reference to its source `mongoc_client_t`. we ended up getting rid of `NextOperation` altogether and now we just run that code via the `Operation.execute` block that takes in a closure and runs it via thread pool.
(this does make me realize that, since that `execute` method bypasses the check for the client being open, methods you call on a cursor created before closing the client are allowed to execute after you've called `MongoClient.close()`. so maybe we need to introduce a check on that path as well...)

going back to your original point, getting rid of the explicit connection case still leaves the session one. can you elaborate on what your concerns are there?
if you think it would be clearer we could move the call to `session.getConnection` outside of the lock code and only call `tryCheckout` in `withLock` if the session is nil.

Reply:
wrt explicit connections: SGTM.
wrt locking: I just want to make sure we aren't complicating our codebase to match the behavior of libmongoc, where we can avoid it. In a future where we don't use libmongoc, we would not want to lock here, and arguably the lock here is guiding behavior of the pool (as you note) so that is most likely better deferred to the pool. As it stands, it might not make a huge difference so I'm fine approving this as-is.